File: src/_322.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    int coinChange(vector<int> &coins, int amount)
    {
        vector<int> dp(amount + 1, -1);
        sort(coins.begin(),coins.end());
        dp[0] = 0;
        int t;
        for (int i = 0; i < amount + 1; ++i)
        {
            t = numeric_limits<int>::max();
            for (auto coin : coins)
            {
                if(i - coin < 0)break;
                if (dp[i-coin] >= 0)
                {
                    t = min(t, dp[i - coin] + 1);
                }
            }
            if(t != numeric_limits<int>::max())
            {
                dp[i] = t;
            }
        }
        return dp[amount];
    }
};

int main()
{
    return 0;
}

File: 322.py
from typing import List
from math import inf

class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [0] + [-1] * amount
        for i in range(1, amount+1):
            f = [1 + dp[i-coin] for coin in coins if i >= coin and dp[i-coin] != -1]
            dp[i] = min(f) if f else -1
        
        return dp[amount]
    
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [0] + [inf] * amount
        for coin in coins:
            for i in range(coin, amount+1):
                dp[i] = min(dp[i], dp[i-coin] + 1)
        return dp[amount] if dp[amount] < inf else -1